[SourceControl] Creates a Xwt Credentials dialog with key selection

netonjm opened the pull request:
When the user connects to a repository using ** git ssh protocol ** and ** does not have an ssh-agent running **, he needs to use an ssh key.

Until now, we were doing 2 steps to authenticate:
Opening a not descriptive FileOpen dialog to select this key file... and before that a second dialog with the user and password fields.

This was not obvious for all the users, and some of them got confused over this step.

To resolve this, we merged both into a single screen in a new XwtCredentialsDialog, which includes additional information error from the key file selected.

What checks do now in this new dialog:

- The key file exists in the path
- if the key has passphrase (is this is true enables the password field)
- if ".pub" file exists with the selected private key file
- Ok button is only enabled in case all the required fields are validated

As @sevoku proposed we revamped native CredentialsDialog in Xwt

Fixes VSTS #675468 - Can't access remote repos over SSH in VS Comm macos
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/675468

New Xwt dialog:
![image](https://user-images.githubusercontent.com/1587480/52857629-0131bb00-3128-11e9-9bea-5bda581b41f8.png)


Old Gtk dialog:

![image](https://user-images.githubusercontent.com/1587480/52829141-df0a4f80-30cb-11e9-8082-a7a1505b8007.png)

Note: I detected after doing some checkouts we get some log errors connecting, probably this issue not related but it would be great to double check

netonjm committed:
[SourceControl] Creates a Xwt Credentials dialog with key selection
Fixes VSTS #675468 - Can't access remote repos over SSH in VS Comm macos

netonjm committed:
[SourceControl] Adds native looking to GitCredentials

netonjm committed:
[SourceControl] Adds a Public key field to Credentials dialog + visual tweaks

Therzok commented on main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/GitCredentials.cs:
```diff
@@ -127,29 +127,28 @@ public static Credentials TryGet (string url, string userFromUrl, SupportedCrede
-						}
+							var engine = Platform.IsMac ? Xwt.Toolkit.NativeEngine : Xwt.Toolkit.CurrentEngine;
+							engine.Invoke (() => {
+								var xwtDialog = new XwtCredentialsDialog (url, types, cred) {
```
We should wrap this in a `using` for code clarity.

Therzok commented on main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/XwtCredentialsDialog.cs:
```diff
@@ -0,0 +1,287 @@
+
+			var passwordLabel = new Label () {
+				TextAlignment = Alignment.End,
+				Text = GettextCatalog.GetString (type == SupportedCredentialTypes.Ssh ? "Passphrase:" : "Password:"),
```
This won't be picked up by the localization crawler.

Therzok commented on main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/XwtCredentialsDialog.cs:
```diff
@@ -0,0 +1,287 @@
+		{
+			var dialog = new Components.SelectFileDialog (GettextCatalog.GetString ("Select a public SSH key to use.")) {
+				ShowHidden = true,
+				CurrentFolder = System.IO.File.Exists (privateKeyLocationTextEntry.Text) ? 
```
NIt: formatting

```
CurrentFolder = System.IO.File.Exists (privateKeyLocationTextEntry.Text)
		? System.IO.Path.GetDirectoryName (privateKeyLocationTextEntry.Text)
		: Environment.GetFolderPath (Environment.SpecialFolder.Personal),
```

Therzok commented on main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/XwtCredentialsDialog.cs:
```diff
@@ -0,0 +1,287 @@
+						passwordEntry.SetFocus ();
+					}
+				} else {
+					if (type == SupportedCredentialTypes.Ssh) {
```
Duplicated code?

sevoku commented on main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/XwtCredentialsDialog.cs:
```diff
@@ -0,0 +1,231 @@
+
+		void KeyLocationButton_Clicked (object sender, EventArgs e)
+		{
+			var dialog = new Components.SelectFileDialog (GettextCatalog.GetString ("Select a private SSH key to use.")) {
```
For the "public" key yes, but private keys usually don't have a file extension or `.ppk` if generated with PuTTY (are they compatible?).

So maybe following filters:
Private Key:
* no extension or `*.ppk` (Default)
* All Files (*.*)

Public Key:
* `*.pub` 
* All Files (*.*)

netonjm replied:
👍

netonjm committed:
[SourceControl] Some minor fixes and styling

netonjm committed:
[SourceControl] Replaces the Error label to use a InformationPopoverWidget
And  includes filtering in FileSelection dialogs

netonjm commented:
![image](https://user-images.githubusercontent.com/1587480/52878805-d6faf000-315d-11e9-867b-3078ca2a6636.png)
Now this is how it looks

sevoku committed:
[Gtk] XwtCredentialsDialog small improvements

sevoku committed:
[Git] Add basic private key validation

rodrmoya commented on main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/GitCredentials.cs:
```diff
@@ -127,29 +127,28 @@ public static Credentials TryGet (string url, string userFromUrl, SupportedCrede
+							engine.Invoke (() => {
+								using (var xwtDialog = new XwtCredentialsDialog (url, types, cred)) {
+									xwtDialog.TransientFor = MessageService.RootWindow;
+									xwtDialog.Show ();
```
This should use the newly added `GetFocusedTopLevelWindow` -> https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Desktop/PlatformService.cs#L510 instead of using RootWindow directly, as that causes problems with the Start Window or when the main window is hidden, which is now possible. And even better, the dialog should be shown via the `MessageService`, as that one has/will have the code (
https://github.com/mono/monodevelop/pull/7169) to parent the dialog correctly.

netonjm replied:
Fantastic contribution rodrigo, thanks for notifying!

sevoku committed:
[Git] CredentialsDialog a11y improvements

sevoku committed:
[Git] Don't stretch credential entries vertically

sevoku committed:
[Git] Improve Credentials dialog layout

sevoku committed:
[Git] Fix Credentials dialog entry events

sevoku committed:
[Git] Fix Credentials dialog command handling

sevoku committed:
[Git] Fix Credentials dialog OnDispose

sevoku committed:
[Git] Refactor XwtCredentialsDialog

sevoku committed:
[Git] Use XwtCredentialsDialog for basic auth

sevoku committed:
[Git] Always fill presets in XwtCredentialsDialog

sevoku committed:
[Git] Improve ssh key selection

* add default id_rsa key to inital keys to try
* take into account that the credentials dialog allows key selection
* support private keys with unusual names

sevoku committed:
[Git] Improve default id_rsa handling when probing keys

Therzok commented:
Is this superseded by https://github.com/mono/monodevelop/pull/7171 ?

sevoku committed:
[Git] Skip public files when scanning for available shh keys

sevoku committed:
[Git] Don't reprompt credentials after cancelling

netonjm commented:
@sevoku  rebased!

sevoku committed:
[Ide] Fix ProgressDialog parenting on native windows

Don't fail if the parent window is native, MessageService
will take care of aligning the dialog.

sevoku committed:
[Git] Fix credentials dialog parent window

sevoku committed:
[Git] Never fail silently on cancelled operation

sevoku committed:
Merge pull request #7171 from mono/update-libgit2sharp-fix-675468-improvements

[Git] Improve new SSH key dialog and selection logic

sevoku commented:
This can be merged now, we'll bring this up in the UX board (cc @jmatthiesen) for feedback (can be still changed for the final 8.0)
